src/Control/FunFlow/Diagram.hs

```diff
@@ -0,0 +1,59 @@
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE InstanceSigs          #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+-- | A Diagram is a representation of an arrow with labels.
+--   No computation is actually performed by a diagram, it just carries
+--   around a description.
+--   We explot the fact that in 7.10,
+--   everything is typeable to label the incoming and outgoing node types.
+module Control.FunFlow.Diagram where
+
+import           Control.Arrow
+import           Control.Arrow.Free (ArrowError (..))
+import           Control.Category
+import           Data.Proxy         (Proxy (..))
+import qualified Data.Text          as T
+import           Prelude            hiding (id, (.))
+
+
+newtype NodeProperties = NodeProperties {
+  labels :: [T.Text]
+}
+
+emptyNodeProperties :: NodeProperties
+emptyNodeProperties = NodeProperties []
+
+data Diagram ex a b where
+  Node :: NodeProperties
+       -> Proxy a
```

Comment:
Do we really need the `Proxy` here - I really don't find this idiomatic haskell. Just leave it out ?

Reply:
We don't - but there's some benefit to having it there, as it means we can get hold of `Typeable a` from the `Node` constructor and thus label the incoming and outgoing types from a computation.